### ISSUE TITLE
Suggestion: Make the hardcoded cluster size user selectable

### ISSUE THREAD

**Issue by mkver** [mbunkus/mkvtoolnix#1703]
Hello,

there are currently three criteria by which mkvmerge decides when a cluster is full in order to begin a new cluster: Based upon time (the default maximum time is 5000ms), based upon the number of blocks in a cluster (default is 65535) and based upon the size of a cluster. The value used is 1500000 B and it is [hardcoded](https://github.com/mbunkus/mkvtoolnix/blob/master/src/merge/cluster_helper.cpp#L145). Why can't this behaviour be changed like the other two? And isn't the default value a bit low? For a 5s cluster it corresponds to a bitrate of (roughly) 2.4Mbit/s so that there will be lots of clusters shorter than 5s where the next cluster does not begin with a keyframe (this is true for ordinary HD content). If I had to choose an arbitrary size limit, I would choose 6 MB: Blu-rays have a maximum AV-bitrate of 48 Mbit/s and a GOP length limit corresponding to one second* and 6 MB entails that every GOP of a remuxed Bluray can be written into one cluster. For a 5s cluster this gives a bitrate limit of 9.6Mbit/s -- this includes a lot of use cases. And the additional 4.5 MB shouldn't be too much for the players either: Today HDDs usually have sequential read speeds of around 100MB/s or more, not to mention SSDs. And even [smartphones](http://www.phonearena.com/news/Android-storage-speed-comparison-which-phone-has-the-fastest-IO-performance_id65588) offer enough sequential read performance for this.
By the way, the only reason why you include this restriction is because if one seeks to a video keyframe (which is of course at the beginning of a new cluster) the needed audio is usually in the cluster before and this cluster needs to be read in full. Is that right? (Before the change that video keyframes are put at the beginning of a new cluster there was another reason: Less data at the beginning of a cluster need to be read and discarded. The [original bug report](https://github.com/mbunkus/mkvtoolnix/issues/500) for this change already refers to the hardcoded size limit which makes me wonder: Is this size limit useless after the improvements that this bug report initiated?)
And I have another question: Think of a file with a video stream with a GOP length of 10s (very common for x264) and one or more audio tracks that has been merged by mkvmerge using default values for cluster-length and cue generation. Then we usually get a 5s cluster starting with a video keyframe, a 5s cluster not starting with a keyframe and then another 5s cluster starting with a keyframe. Assume I want to seek to the second keyframe and that the audio that should be played alongside said keyframe is in the second cluster. Then the player needs to seek to this cluster, but said cluster doesn't have any cues (because by default cues are only written if a cluster contains a video keyframe). Does this mean that the player has to read the complete first and second clusters? If the answer to this question is yes, then what's the point of splitting the first GOP into two clusters? If mkvmerge continues to add cues only for keyframes then one should rethink whether one needs the size and time limits for clusters at all. Creating a new cluster with a new keyframe should be sufficient.

Grüße
Andi

*: As far as I know, there is also a second profile with a much lower maximum bitrate (15MBit/s ?) where a GOP length of 2s is allowed. But that's unimportant.
PS: The German translation contains a typo: The balloon text for the NALU-Größenfeldlänge refers to a size limit of 6535 bytes.

**Comment by mbunkus** [comment]
I do appreciate your report, but I'm sorry, I just don't care enough about this topic to invest time in it. Making things configurable is always quite a lot of work (not insurmountably so, but still): adding an option to mkvmerge's help output, parsing the value, error handling, testing, investigate existing test cases if their behavior differs, add documentation, think about how to translate that option into controls in the GUI, adjust the dialogs, implement the functionality, translate all new strings. I also don't think that this would be a setting a lot of users would actually care to play around with.

**Comment by mkver** [comment]
I do understand why you don't want to add another option for the user. But what I do not understand is why the size and length limits exist at all any more. What are they good for (if I may ask)? If I remux using mkclean (from the matroska site) it adds a field about the size of the cluster before this cluster to every cluster (except the first one, of course; and only when muxing without the --unsafe option). This is information that could be used for seeking in my 10s GOP length example (I don't know if players make use of it; probably not). With this information, dividing this GOP into two clusters might seem sensible. (But I still see putting one GOP into one cluster as the better option.) But without it like mkvmerge?

**Comment by mbunkus** [comment]
The limits exist to limit the amount of data lost if a data corruption occurs and the player has to re-sync to the next cluster start.

**Comment by mkver** [comment]
Thanks for the explanation. I did not take into account data corruption or at least I underestimated its importance to you (after all mkvmerge doesn't write those crc32 values). Maybe in the future I will try a custom build for me and my special wishes (that I of course will not republish so that nobody will confuse it with yours).
By the way, you have seem to have forgotten the "PS" above. I will open a separate issue for it.
